fxrunner/src/lib/taskcluster.rs

```diff
@@ -0,0 +1,297 @@
+use tokio::prelude::*;
+
+/// The name of the artifact containing the result of a build job.
+pub const BUILD_ARTIFACT_NAME: &str = "public/build/target.zip";
+
+#[derive(Deserialize, Serialize)]
+pub struct Artifact {
+    pub name: String,
+    pub expires: DateTime<Utc>,
+}
+
+/// An error from Taskcluster
+#[derive(Debug, Display)]
+pub enum TaskclusterError {
+    /// An
+    #[display(fmt = "IO error: {}", _0)]
+    Io(io::Error),
+
+    #[display(fmt = "could not parse URL: {}", _0)]
+    UrlParse(url::ParseError),
+
+    #[display(fmt = "could not list artifacts: {}", _0)]
+    ListArtifacts(reqwest::Error),
+
+    #[display(fmt = "artifact expired on {}", _0)]
+    Expired(DateTime<Utc>),
+
+    #[display(fmt = "no build artifact found")]
+    NotFound,
+
+    #[display(fmt = "an error occurred while downloading the artifact: {}", _0)]
+    DownloadArtifact(reqwest::Error),
+}
+
+impl From<io::Error> for TaskclusterError {
+    fn from(e: io::Error) -> Self {
+        TaskclusterError::Io(e)
+    }
+}
+
+impl From<url::ParseError> for TaskclusterError {
+    fn from(e: url::ParseError) -> Self {
+        TaskclusterError::UrlParse(e)
+    }
+}
+
+impl Error for TaskclusterError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            TaskclusterError::Io(ref e) => Some(e),
+            TaskclusterError::UrlParse(ref e) => Some(e),
+            TaskclusterError::ListArtifacts(ref e) => Some(e),
+            TaskclusterError::DownloadArtifact(ref e) => Some(e),
+            TaskclusterError::Expired(..) => None,
+            TaskclusterError::NotFound => None,
+        }
+    }
+}
+
+/// An API client to download Taskcluster build artifacts.
+pub struct Taskcluster {
+    /// The reqwest Client used for all requests.
+    client: Client,
+
+    /// The URL for the Taskcluster Queue API.
+    queue_url: Url,
+}
+
+impl Default for Taskcluster {
+    fn default() -> Self {
+        Taskcluster {
+            queue_url: Url::parse("https://firefox-ci-tc.services.mozilla.com/api/queue/v1/")
+                .unwrap(),
+            client: Client::new(),
+        }
+    }
+}
+
+impl Taskcluster {
+    pub fn with_queue_url(queue_url: Url) -> Self {
+        Taskcluster {
+            client: Client::new(),
+            queue_url,
+        }
+    }
+
+    /// Download the build artifact from a Taskcluster task.
+    pub async fn download_build_artifact(
+        &mut self,
+        task_id: &str,
+        download_dir: &Path,
+    ) -> Result<PathBuf, TaskclusterError> {
+        let artifacts = self.list_artifacts(task_id).await?;
+        for artifact in artifacts {
+            if artifact.name == BUILD_ARTIFACT_NAME {
+                if artifact.expires <= Utc::now() {
+                    return Err(TaskclusterError::Expired(artifact.expires));
+                }
+
+                let path = download_dir.join("firefox.zip");
+
+                self.download_artifact(task_id, &artifact.name, &path)
```

Comment:
I'm slightly confused - why do we need to loop through the artifacts to find one whose name equals `BUILD_ARTIFACT_NAME`, if the only thing we're going to use of that artifact is its name? Couldn't we just always use `BUILD_ARTIFACT_NAME` in `download_artifact`? We can check the expired property, but I assume that `download_artifact` would fail anyway if the artifact is expired, no?

Reply:
I believe this was only to determine if it was expired or not. I guess a 404 will imply that so I'm happy to change it